### PR TITLE
perf(scope-resolution): stop re-scanning the ParsedFile store once per language

### PR DESCRIPTION
**Pull request by magyargergo**
Follow-up to #3209, which closed FTS and left `scopeResolution` (14.7s, 44% of the edit loop) as the largest remaining target. This breaks it down and takes the one contained win in it.

## The breakdown

`PROF_SCOPE_RESOLUTION=1` already existed and reports the internal split; marks around the phase supply the rest.

| step | ms | share |
| --- | ---: | ---: |
| ParsedFile store rehydration | 4426 | 30% |
| **`emit`** | **7161** | **49%** |
| `resolve` | 966 | 7% |
| `finalize` | 552 | 4% |
| `extract` | 369 | 3% |
| teardown, misc | ~750 | 5% |

`extract` is small because the parse cache works — 2121/2121 pre-extracted hits on the TypeScript pass. Nothing re-parses.

## What this PR fixes

Scope resolution calls `loadParsedFilesForPaths` once per language, and every call walks **every** shard in the store. The skip decision needs the envelope's path listing, and that listing is only trustworthy after the payload digest is checked — so a pass that wants 50 Python files still opens and SHA-256s all 413 shards / 301MB of a TypeScript-dominated store to prove it can skip them. **A pass wanting a single file costs 335ms.** The cost scales with *language count*, not with how many files that language has, so a polyglot repo pays it worst.

`tryLoadV8Cache` now returns the listing it already parsed for that skip decision, and the store memoizes it per run.

```
before   python 411ms   typescript 2872ms   javascript 507ms  = 3834ms
after    python 415ms   typescript 2805ms   javascript 248ms  = 3484ms
```

min-of-5. −350ms here, roughly −250ms per additional language elsewhere. The first pass is unchanged by construction — it is what populates the memo.

## The failure mode, and the test for it

A listing memo can produce a **false skip**: a pass concludes a shard holds nothing it wants, and those files silently never reach the graph. That is an exit-0 wrong answer, not a crash. The new test walks four passes with disjoint wants over one store, plus a shard written after the memo is warm. Forcing the skip fails it (verified by mutation).

Keyed on size+mtime as well as name. Shard names are content-addressed so a collision should be impossible, but that invariant lives in the parse-cache keying rather than here, and one `stat` per shard is a few ms against the hundreds this saves. The memo holds one store directory at a time, so a new repo in a long-lived MCP process drops the previous set instead of accumulating.

## What is a floor, and what is next

The digest is **not** the cost: SHA-256 over all 301MB takes 134ms (2.25 GB/s, hardware-accelerated), so swapping to CRC-32 buys ~90ms and costs an envelope-format bump. Not worth it. The rest is `v8.deserialize` (~1240ms) plus the cross-shard intern walk (~1085ms), and the intern walk is not optional — dropping it regresses retained heap ~59%, which is #2649's constraint.

`emit` is the real target — 7.2s, ~21% of the edit loop, spread across a fan of passes with no hot inner loop and nothing quadratic:

```
1828ms  emitCallableValueFlow      480ms  emitFreeCallFallback
1590ms  emitReceiverBoundCalls     426ms  emitReferencesViaLookup
 681ms  resolveDefGraphId          313ms  emitUniqueNamePropertyAccesses
 529ms  lookupCore                 280ms  emitReturnShapeMemberAccesses
 472ms  tryEmitEdge                222ms  emitPropertyDispatchCalls
```

Micro-optimizing any of them is not the win. A one-file edit re-emits all 163,094 edges to write a 3,980-node subgraph, because the pipeline rebuilds the full in-memory graph every run and only the DB *write* is incremental. Making `emit` incremental is a design, not a patch.

## Verification

`tsc --noEmit` clean. 112 tests pass across `parsedfile-store` / `v8-sidecar` / `sidecar-recovery`; 127 including the warm-cache coverage and parse-impl suites. End-to-end on a true incremental run the graph is identical: 51,288 nodes / 163,094 edges / 2106 clusters / 759 flows. The other `tryLoadV8Cache` caller (`parse-cache.ts`) reads only `.kind`/`.value`, so the added `paths` field is inert there.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

## Summary

*This appears to be a storage and analysis-pipeline performance change, centered on parsed-file loading and V8 cache handling. The reach is critical despite low per-file risk, so review should focus on behavior shared across the storage and ingestion paths.*

**🔴 CRITICAL blast radius. A storage-layer performance change in `gitnexus/src/storage/parsedfile-store.ts` and `gitnexus/src/storage/v8-sidecar.ts` reaches 19 dependents across direct and transitive graph paths.**

The primary implementation changes are in `loadParsedFilesForPaths`, `maybeYieldAndGc`, and `sanitizeCallableFlowSites`, alongside `readExact`, `tryLoadV8Cache`, and `writeV8CacheFile`. The accompanying benchmark documentation in `gitnexus/bench/analyze-phase-breakdown.md` frames the change around scope-resolution performance, while `gitnexus/test/unit/parsedfile-store.test.ts` is the relevant validation surface.

Impact is concentrated in `Storage`, with additional connections to `Pipeline-phases`, `Tree-sitter`, `Ingestion`, `Pipeline`, and `Lbug`. Review the parsed-file enumeration and cache read/write paths first, then verify that the unit coverage preserves the expected loading, yielding, flow-site sanitization, and V8 cache behavior across the affected execution flows.

_Added by GitNexus for PR #3211. Edit freely — this block is replaced on the next review, everything above it is left untouched._

### CODE REVIEW

**Review by koriyoshi2041 — APPROVED** [review]
Validated exact head `94a75663`. The authenticated listing is cached only after the sidecar digest/path-count checks, unknown or changed shard identities fall back to a full read, and the shard roster is still refreshed per call. Focused storage/sidecar coverage passed locally (112/112), including disjoint per-language wants and a shard added after memo warm-up; `npx tsc --noEmit` also passed. The reported hosted test failure is the platform-sensitive Windows wrapper timeout with 20,993 tests passing and 0 assertion failures, not a changed-path failure.

**Review by magyargergo — COMMENTED** [review]
# PR #3211 tri-review digest

**PR:** [perf(scope-resolution): stop re-scanning the ParsedFile store once per language](https://github.com/abhigyanpatwari/GitNexus/pull/3211)  
**HEAD:** `9df1de0e14c9afad383c2891989e44e8e8f7b9a4` · **BASE:** `d1463977c8804dbf71d5aeb3cb949c5ba37cc828`  
**Surface:** 4 files (`parsedfile-store.ts`, `v8-sidecar.ts`, `parsedfile-store.test.ts`, `analyze-phase-breakdown.md`)

---

## 1. Engine breakdown + liveness

| Engine | Method | Status |
|---|---|---|
| **A — Claude** | GitNexus swarm + ce-code-review | **Partially degraded.** Typed Task spawn failed for facts, risk, security, docs, and CE testing / maintainability / project-standards / performance / reliability. Those artifacts are **coordinator-fallback** (one coordinator voice), **not** independent Claude-engine confirmation. |
| **A — Claude (independent)** | Typed agents that succeeded | `claude-swarm-hygiene`, `claude-swarm-test-ci`, `claude-ce-correctness`, `claude-ce-adversarial` |
| **B — Codex GPT-5.5 / xhigh** | Swarm + CE + adversarial | **All three legs live** (not skill-not-found). B1 first attempt hit model capacity (`codex-swarm.attempt1-capacity.md`); retry produced a real Solo `gitnexus-pr-review`. |

**Live Codex legs**

- **B1** `codex-swarm.md` — Solo gitnexus-pr-review (all 6 review personas; critic skipped by contract)
- **B2** `codex-ce.md` — ce-code-review report-only (correctness, testing, maintainability, performance, project-standards, adversarial)
- **B3** `codex-adversarial.md` — native `adversarial-review`, **verdict: approve**, no material findings

Do **not** treat coordinator-fallback + Codex as Strong A/B. Strong cross-engine is only independent Claude ∩ Codex. Fallback domains (facts / risk / security / docs / several CE personas) are **not** a full two-engine swarm+CE A/B. Codex was **not** unavailable.

---

## 2. What the reviews validated

Independent Claude correctness + adversarial, Codex B2 correctness/adversarial, and Codex B3 all **failed to construct a production false-skip**. The named failure mode (memoized listing omits a path the shard still holds → exit-0 missing graph files) has **no concrete analyze-path trigger**.

Code-read confirmation of the fail-closed design:

- Skip predicate is the same intersection `tryLoadV8Cache` already used; memo stores only a listing that pass already authenticated (`v8-sidecar.ts:315-333`, `parsedfile-store.ts:322-338`).
- Missing/invalid listing, `pathCount` mismatch, empty listing, digest miss, or `stat` failure → no identity / no `paths` → no memo write → full reopen. Empty `wantPaths` returns before the loop (`parsedfile-store.ts:291`).
- `V8_CACHE_FORMAT` stays **5**. Old shards with `pathBytes === 0` never memoize.
- `parse-cache.ts:1029-1032` omits `wantPaths`, reads only `kind === 'hit'` + `.value`; added `paths` is inert.
- `clearParsedFileStore` wipes the dir **and** `forgetShardListings()` (`parsedfile-store.ts:170-172`).
- Concurrent two-dir MCP: `memoFor` replaces the singleton; the in-flight call keeps its local `Map`; the next call for the dropped dir starts empty → re-read, not a cross-repo skip (`parsedfile-store.ts:276-278`).
- New shard after warm memo is found because the roster is `readdir` per call (unit test `e.c` path).
- Production persist finishes before the first `loadParsedFilesForPaths` (parse-impl `:1828` then `:1840`; scope-resolution `phase.ts:404-407`). Writers are tmp+rename.
- Scope-resolution miss still falls back to source read + `extractParsedFile`. The parse-impl route pair has **no** extract fallback, but that would matter only if a listing were stale — not constructed.

---

## 3. Headline

**No production correctness / security blocker found.** The only postable, rule-3-verified item is a **P2 test-strength gap**: the new memo test locks *wrong-memo* (missing keys) but would still pass if the memo never skipped.

Pending CI and required-reviewer merge state are **process**, not code defects. No failing check on the live re-check.

---

## 4. Inline-candidate findings (rule 3 cleared)

### P2 — New memo test does not fail if the memo never skips

- **path:** `gitnexus/test/unit/parsedfile-store.test.ts`
- **line:** 109 (RIGHT; added hunk)
- **severity:** P2
- **problem:** Five loads assert returned key sets only. That is a *wrong-listing* net, not a *memo-present* net.
- **trigger:** Persist `chunk-0/1/2` with `a.c`–`d.c`. Call `loadParsedFilesForPaths` for `{a.c}`, then `{c.c,d.c}`, then `{b.c}`, then the full set. If `tryLoadV8Cache` stopped returning `paths`, or `listings.set` never ran (`parsedfile-store.ts:336-337`), later passes would still open + deserialize intersecting shards and return the same keys. `expect([...first.keys()])` at `:109-112` and the `e.c` key assert at `:118` would still pass. Pre-PR `wantPaths` filtering already produced those key sets.
- **evidence:** `[code-read]` `parsedfile-store.test.ts:104-118` (key-only expects); `parsedfile-store.ts:325-338` (memo skip vs `tryLoadV8Cache` fallback).
- **fix:** On a later pass whose `wantPaths` miss a memoized listing, spy `v8.deserialize` / `fs.open` (or assert `tryLoadV8Cache` is not re-entered for that shard). Also assert `tryLoadV8Cache` returns authenticated `paths` on hit and skip.
- **corroboration:** **Strong** — independent Claude `swarm-test-ci` F1 **and** Codex B2 testing (`parsedfile-store.test.ts:109`). Independent Claude correctness listed the same item as a testing gap (not a production finding). Coordinator-fallback risk/testing agreed; that is **not** a second Claude engine.

---

## 5. Lower-priority / body-only

| Item | Sev | Weight | Why not inline |
|---|---|---|---|
| Same-path rewrite / size+mtime identity collision untested | P2 gap | Independent Claude test-ci F2; Codex B2 residual; correctness/adversarial **refuted as a live production sequence** | P2 without a constructed production trigger; adversarial/correctness dropped the bug. Gap only. |
| `forgetShardListings` untested (`clear` test only checks empty load after `rm`) | P2 gap | Independent Claude test-ci F3 + correctness testing-gap; Codex did not elevate | P2, no second-engine finding |
| `v8-sidecar.test.ts:51` asserts `kind === 'skip'` only, not `paths === ['a.c']` | P2/P3 | Independent Claude test-ci F4; Codex B2 residual; coordinator-fallback P3 | **Unchanged file** — cannot anchor on the three-dot RIGHT side. Typecheck already requires `paths` on `V8CacheSkip`. |
| Process-global one-dir memo untested | P3 | Independent Claude test-ci F5 | Weak; fail-closed (re-read) per correctness/adversarial/B3 |
| parse-cache caller not covered for new `paths` | P3 | Independent Claude test-ci F6 | Weak; skip branch unreachable without `wantPaths` |
| Bench table `4426ms` rehydration vs `= 3834ms` three-pass sum (`analyze-phase-breakdown.md:179` / `:201`) | P3 | Codex B1 only | Weak. Plausible different scopes (phase-split share vs min-of-5 load timings). Do not treat as a confirmed numeric error. |
| Three merge-from-main commits; approval cites feature SHA | P3 | Independent Claude hygiene F1/F2 | History / GitHub process. Three-dot patch-id matches the feature; rebase not required. |
| Pending current-head CI (Codex B1 called this P2 merge-block) | process | B1 + hygiene F3 + test-ci F7 | **Not a code defect.** See §8. |

---

## 6. Refuted suspicions (excluded)

Probed by independent Claude correctness (H1–H9), independent Claude adversarial (hunts 1–7), Codex B2 correctness/adversarial, and Codex B3 (approve). **No divergence found.**

- Memo stores the wrong shard’s listing (identity is abs path + size + mtime).
- Concurrent two-repo MCP poisons listings (local `Map` + dir swap → reload).
- Shard written after memo is skipped (roster re-read; `e.c` test).
- Memo skip vs `tryLoadV8Cache` skip diverge on path spelling (same `wantPaths.has`).
- Listing is a subset of the deserialized payload (writers pass `shardFilePaths` of the same array; newline/CR/NUL → no listing → no memo).
- parse-cache breaks on required `V8CacheSkip.paths` (no `wantPaths` → no skip).
- `clearParsedFileStore` forget vs persist-without-clear leaves a false skip.
- Same-path overwrite after memo with preserved size+mtime on the analyze path (no persist-after-first-load; tmp+rename). Residual only for `cp -p` / `utimes` / coarse-mtime FS — **not** a current caller.
- Implicit envelope / `PARSE_CACHE_VERSION` bump (`V8_CACHE_FORMAT` still 5).
- Memo as integrity bypass or cross-repo secret leak (coordinator-fallback security; listings are repo-relative paths).
- Prior-context probe class #52589 (silent skip/omit) — same *shape*, not transplanted; no trigger here.

---

## 7. Test gaps (consolidated)

1. **Memo-present assertion missing** — see §4 (inline).
2. Warm memo → rewrite same `shardId` with different paths (force old size+mtime) → must not skip.
3. Load → `clearParsedFileStore` → persist same names, different paths → must not reuse listings (`forget` currently untested).
4. Invalid/empty listing on pass 1 must not be memoized (second pass still deserializes).
5. Direct `tryLoadV8Cache` `paths` asserts on hit/skip (`v8-sidecar.test.ts` unchanged).
6. No parse-impl fixture for data-route `wantPaths` then import-target `wantPaths` (first production memo *use*, and the only load pair without `extractParsedFile` fallback).
7. File is not in `cross-platform-tests.ts`; Windows/macOS platform shards will not run it. Ubuntu `vitest run --shard` will, once coverage 1/3 finishes.

---

## 8. CI status + required classifications

**Merge state:** `review blocked` (also `checks pending` on a re-queued run; hygiene’s durable class is required review — `reviewDecision` empty, `azizur100389` still requested; `koriyoshi2041` APPROVED on HEAD `9df1de0e`).  
**Branch hygiene:** `merge-from-main commit present but harmless and merge-safe`  
**Final verdict:** `production-ready with minor follow-ups`

Coordinator re-check (`gh pr checks 3211`) after a **new** Actions run was queued (`34203644*`): **no failures**. Currently **pass:** Vercel + Vercel Preview Comments. **Pending:** Analyze (js/ts, python), autofix, gitleaks, review, Build & Push gitnexus, Build & Push gitnexus-web. Ubuntu coverage shards are **not in the rollup yet** (re-queue). An earlier synthesis snapshot that said coverage 2/3 and 3/3 had passed is **stale** — do not treat that as HEAD evidence.

Codex B1’s P2 “required validation has not settled” remains accurate as **process**. Not a code finding.

---

## 9. Coverage limits

- Four-file diff was read in full; no huge-PR partial-scope disclaimer needed for *files*.
- Engine A swarm/CE A/B is **incomplete** for fallback domains (see §1).
- GitNexus MCP/`impact` against this worktree was **UNKNOWN** (LadybugDB file v43 vs build v42; workspace index is main, not HEAD). Production callers recovered by text: `phase.ts:404-407`, `parse-impl.ts:1828` / `:1840`. File-level bot CRITICAL/19 dependents is **reach**, not a constructed bug. Symbol-level LOW is **unresolved**, not unused.
- Review lanes did not execute the new vitest (read-only / symlink `node_modules`).
- Prior-context (`#3209`, `#2649`, `#51098`, `#52589`) used only as probes; none became findings.

---

## 10. Close

Automated multi-tool digest (Claude independent lanes + coordinator-fallback + Codex GPT-5.5 swarm/CE/adversarial). Verify in the worktree before acting. No P0/P1. One P2 test-strength inline. Do not treat CI as green until the re-queued coverage shards and remaining jobs finish.

---

File: gitnexus/test/unit/parsedfile-store.test.ts [hunk 1]
```diff
@@ -86,6 +86,41 @@ describe('parsedfile-store', () => {
...
+      await persistParsedFileChunk(dir, 'chunk-0', [makeParsedFile('a.c'), makeParsedFile('b.c')]);
+      await persistParsedFileChunk(dir, 'chunk-1', [makeParsedFile('c.c')]);
+      await persistParsedFileChunk(dir, 'chunk-2', [makeParsedFile('d.c')]);
+
+      const first = await loadParsedFilesForPaths(dir, new Set(['a.c']));
+      const second = await loadParsedFilesForPaths(dir, new Set(['c.c', 'd.c']));
+      const third = await loadParsedFilesForPaths(dir, new Set(['b.c']));
+      const fourth = await loadParsedFilesForPaths(dir, new Set(['a.c', 'b.c', 'c.c', 'd.c']));
+
+      expect([...first.keys()].sort()).toEqual(['a.c']);
```
Comment (magyargergo) [inline gitnexus/test/unit/parsedfile-store.test.ts]: **P2** — This test would still pass if the shard-listing memo never skipped.

Trigger: persist `chunk-0/1/2`, then four `loadParsedFilesForPaths` calls with disjoint wantSets. If `tryLoadV8Cache` stopped returning `paths` (or `listings.set` at `parsedfile-store.ts:336-337` never ran), later passes would still open + deserialize intersecting shards and return the same keys. These expects only check key sets — the same net the pre-PR `wantPaths` filter already provided. The `e.c` case proves roster refresh, not that a memoized listing avoided `open`/`v8.deserialize`.

Evidence: `:104-118` key-only asserts; skip path is `parsedfile-store.ts:325-328`. `[code-read]`

Fix: on a later pass whose `wantPaths` miss a memoized listing, spy `v8.deserialize` / `fs.open` (must not run for that shard). Optionally assert `tryLoadV8Cache` hit/skip includes authenticated `paths`.

Engines: Claude `swarm-test-ci` + Codex B2 testing (Strong). Claude correctness recorded the same gap; not a production false-skip.

Comment (magyargergo) [reply gitnexus/test/unit/parsedfile-store.test.ts]: > This test would still pass if the shard-listing memo never skipped.

Addressed: later passes now spy `fs.open` and `v8.deserialize`. A miss against a memoized listing must not reopen that shard — envelope skip still opens; only the memo skip does not. Landed in d25fccd1f.

**Comment by magyargergo** [comment]
> The only postable, rule-3-verified item is a **P2 test-strength gap**: the new memo test locks *wrong-memo* (missing keys) but would still pass if the memo never skipped.

Addressed in d25fccd1f — the shard-listing memo test now spies `fs.open` and `v8.deserialize` on later passes. A miss against a memoized listing must not reopen that shard (envelope skip still opens; only the memo skip does not). The inline thread on `parsedfile-store.test.ts` is the same item.

**Review by koriyoshi2041 — APPROVED** [review]
Re-reviewed `ad05236d`. The updated test now proves the second lookup avoids both reopening and deserializing the shard, so the memo fast path is observable rather than inferred from equal keys. The listing remains digest-authenticated before memoization, identity changes invalidate reuse, and store clearing drops the run-local memo.

The current macOS failure is in `analyze-index-lock-concurrency.test.ts`, outside this PR’s four-file surface; the other platform shards and changed-path gates are green.
